### ISSUE TITLE
JSON Standard Library, requiring ext.rb causes a load error

### ISSUE THREAD

**Issue by RJSent** [oracle/truffleruby#2032]
Tested on Manjaro Linux, TruffleRuby 20.1.0 installed via Rbenv

I was using a gem that had the Multi_Json gem as a dependency. Multi_Json requires 'json/ext'  in their json_gem.rb file, which made me first notice this.

```
$ ruby --version
truffleruby 20.1.0, like ruby 2.6.5, GraalVM CE Native [x86_64-linux]
$ ruby -e "require 'json/ext'"
/home/richard/.rbenv/versions/truffleruby-20.1.0/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `raise_load_error': cannot load such file -- json/ext/parser (LoadError)
        from /home/richard/.rbenv/versions/truffleruby-20.1.0/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `gem_original_require'
        from /home/richard/.rbenv/versions/truffleruby-20.1.0/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `require'
        from /home/richard/.rbenv/versions/truffleruby-20.1.0/lib/json/lib/json/ext.rb:7:in `call'
        from /home/richard/.rbenv/versions/truffleruby-20.1.0/lib/json/lib/json/ext.rb:7:in `require'
        from /home/richard/.rbenv/versions/truffleruby-20.1.0/lib/json/lib/json/ext.rb:7:in `Ext'
        from /home/richard/.rbenv/versions/truffleruby-20.1.0/lib/json/lib/json/ext.rb:6:in `JSON'
        from /home/richard/.rbenv/versions/truffleruby-20.1.0/lib/json/lib/json/ext.rb:3:in `<top (required)>'
        from -e:1:in `require'
        from -e:1:in `<main>'
```

A similar load error occurs if I comment out `require 'json/ext/parser'`, continue to line 8, and attempt to `require 'json/ext/generator'`.

**Comment by eregon** [comment]
We currently ship json/pure in TruffleRuby (but the json C extension can also be installed as a gem if desired).
Is there an actual issue in practice?

This works fine for me:
```
$ gem i multi_json
Fetching multi_json-1.14.1.gem
Successfully installed multi_json-1.14.1
1 gem installed

$ gem list | grep json                                                 
json (default: 2.1.0)

$ ruby -e 'require "multi_json"; p MultiJson.load(%{{"abc":"def"}})'
{"abc"=>"def"}
```

Using `-d` as in `ruby -d -e 'require "multi_json"; p MultiJson.load(%{{"abc":"def"}})'` will print many exceptions (which are caught, notably [here](https://github.com/intridea/multi_json/blob/66864817eb2fae85e9a93d8823877d75851ad666/lib/multi_json.rb#L57)), but so does it on MRI.

So I guess the only issue is if anyone requires `json/ext` explicitly, does not rescue the `LoadError`, and also does not install the `json` gem. Is there such a case? I don't see the issue with `multi_json` at least.

**Comment by eregon** [comment]
I found an issue, if `json` is required first:
```
$ ruby -d -rjson -rmulti_json -e 'p MultiJson;  p MultiJson.load(%{{"abc":"def"}})'
Using Pure library for JSON.
...
Exception: `LoadError' /home/eregon/.rubies/truffleruby-dev/lib/mri/rubygems/core_ext/kernel_require.rb:132:in `require' - cannot load such file -- json/ext/parser
Exception: `MultiJson::AdapterError' /home/eregon/.rubies/truffleruby-dev/lib/gems/gems/multi_json-1.14.1/lib/multi_json.rb:110:in `load_adapter' - Did not recognize your adapter specification (cannot load such file -- json/ext/parser).
```

Due to MultiJson [assuming](https://github.com/intridea/multi_json/blob/66864817eb2fae85e9a93d8823877d75851ad666/lib/multi_json.rb#L50) that `defined?(::JSON::JSON_LOADED)` means the C extension available, which is not the case, since `json/pure` also defines that constant.

Is that the issue you have?

**Comment by eregon** [comment]
@RJSent was this solved in your case? Otherwise I think we should leave the issue open, and probably do a PR to multijson or see what can be done on TruffleRuby's side.

**Comment by RJSent** [comment]
@eregon Sorry, it wasn't solved yet. I'll reopen the issue.

**Comment by eregon** [comment]
Issue filed: https://github.com/intridea/multi_json/issues/196

**Comment by eregon** [comment]
PR to fix this in `multi_json`: https://github.com/intridea/multi_json/pull/197

**Comment by eregon** [comment]
I'm unsure if we can address this issue on TruffleRuby's side.
`json/ext.rb` exists in the `json_pure` gem too, and removing it would not help for this case.

I think the only way to fix it on TruffleRuby's side would be to have a patch for `multi_json.rb` to apply that change by redefining `MultiJSON.default_adapter` on older versions, but I am unsure if it's worth it.
Let's see if the PR is merged and there is a release soon, that is the best fix.

**Comment by eregon** [comment]
The fix is now merged in multi_json and released as [multi_json 0.15.0](https://github.com/intridea/multi_json/pull/197#issuecomment-656483113).
Using multi_json 0.15.0 the commands above work.
